### DISCUSSION TITLE
Controllers for .antlers-Files

### DISCUSSION THREAD

**Discussion by nadlig123** [statamic/cms#7597] [Q&A]
I could not quite figure out how to use controllers & the advantages of .antlers-files. How can I pass the data from my collections and data i fetched from an external api-call within the controller both at once to a antlers-view. 

So far I tried:

- Setting up the route  in **web.php** as follows Route::get('home', [HomeController::class, 'home']);
- Having a antlers-view **home.antlers.php** with some content
- and a controller **HomeController**  with following functionality:
 public function home() {
        $data = ***external Api-Call***
        $entry = \Statamic\Facades\Entry::find("home");
        return view('pages/home', $entry ,$data);
  }

But now my view does not include my specified-layout and all in all it seems quite messy. Is there a better way of not having to use individual routing but passing the fetched data from the controller to my antlers-view.

**Comment by jasonvarga** [comment]
When using the standard `view()` helper, you only get the view you asked for, and don't get the template-into-layout behavior other Statamic pages provide.

You should instead use our View class.

```php
use Statamic\View\View;

return View::make('pages/home', $data)->layout('layout');
```

If you don't want to use a controller, you could consider a view model. https://statamic.dev/view-models

In your `home.md` entry, for example, you could add:

```yml
view_model: 'App\ViewModels\HomeViewModel'
```

and then create the view model php file in `app/ViewModels/HomeViewModel.php`

```php
<?php
 
namespace App\ViewModels;
 
use Statamic\View\ViewModel;
 
class HomeViewModel extends ViewModel
{
    public function data(): array
    { 
        return $this->makeApiCall();
    }
}
```

**Reply by nadlig123** [reply]
Thanks. I will try and give a update :D

**Comment by hansipete** [comment]
Where would that ArticleStats php file live? `app/???/ArticleStats.php`

**Reply by jasonvarga** [reply]
Sorry that was a copy/paste mistake. I've updated my comment to rename the class and show the location of the file.

**Reply by hansipete** [reply]
Thank you!
